moor/lib/src/runtime/data_class.dart

```diff
@@ -125,6 +125,17 @@ class Value<T> {
 
   @override
   String toString() => present ? 'Value($value)' : 'Value.absent()';
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Value &&
+          runtimeType == other.runtimeType &&
```

Comment:
I'm not sure if we should always consider the `runtimeType` here. For instance, one might expect `Value<bool>.absent()` to be equal to `Value<String>.absent()` since they have identical behavior. I'm not so sure about `Value<String>(null) == Value<bool>(null)` though.

The `runtimeType` comparison is commonly used to guard against unexpected subclasses messing up equality, but I think we don't need that in this case. Maybe we can change this to `(!present || other.runtimeType == runtimeType)`? I don't really have a strong opinion on this, maybe something else makes more sense.

Reply:
Thanks for the suggestions!

I don't know what the most appropriate way is either, but I agree with your argumentation in principle. I still see the problem that the hash code is the same in this case:
```dart
Value<String>(null).hashCode == Value<bool>(null).hashCode`
```
but `equals` in this case is `false`. 
As far as I can see, there are two possibilities: 
1. We remove `(!present || other.runtimeType == runtimeType)` 
2. We adjust the `hashCode` so that the generic `T` matters, e.g. `(present ? T.hashCode : 0) ^ present.hashCode ^ value.hashCode;` (haven't tested it yet)

I decided to go for the first version first, because it's simpler. But I don't have a strong opinion and I'll adjust it if you like.

I also noticed that in the test `companions support hash and equals` the line
```dart 
expect(first, equals(first));
```
doesn't make sense and I adjusted it accordingly.

Reply:
I think it's ok for `hashCode` to be less precise than `==`. But just using `value` and `present` looks good to me as well, I think we can keep it like that.

However, I'd really like to a test case to ensure `Value.absent() != Value(null)`.

Reply:
Done 👍